tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc

```diff
@@ -387,11 +418,34 @@ void CaptureConvGraphRecursive(HloInstruction* instr,
   final_instr = instr;
 
   HloInstruction *op, *operand0, *operand1;
+  auto fuse_amax = [&]() -> bool {
+    HloComputation* reduce_comp = op->to_apply();
+    HloInstruction* reduce_comp_root = reduce_comp->root_instruction();
+    if (ShapeUtil::IsScalar(op->shape()) &&
+        ShapeUtil::IsScalar(op->operand(1)->shape()) &&
+        op->operand(1)->IsConstant() &&
+        op->operand(1)->literal().GetAsDouble({}) <= 0. &&
+        reduce_comp_root->opcode() == HloOpcode::kMaximum &&
+        reduce_comp_root->operand(0)->opcode() == HloOpcode::kParameter &&
+        reduce_comp_root->operand(1)->opcode() == HloOpcode::kParameter) {
+      aux_outputs.emplace_back(op);
+      graph_string.AppendOp("amax", op, {operand0});
+      return true;
+    }
+    return false;
+  };
+
+  // Copy the current state in case fusion will be unsuccessful or unfavorable.
+  GraphString init_graph_string = graph_string;
+  std::vector<HloInstruction*> init_operands = operands,
+                               init_aux_outputs = aux_outputs;
+  int linear_users = 0, nonlinear_users = 0;
```

Comment:
Call this num_linear_users and num_nonlinear_users and comment what these are, e.g.

> We add each user of instruction to GraphString if possible. For most types of ops, we recrusively call CaptureConvGraphRecursive so that we also add users of the user to the graph. We only make a recursive call for at most one user, and so a linear chain of ops are formed. We count the number of such recursive calls in "num_linear_users" and will not do the rewrite if there is more than 1 linear user. For a user representing an amax calculation, we do not make a recursive call because we don't anticipate a need to fuse users of the amax op into the cuDNN graph. The number of such users is stored in "num_nonlinear_users".

I wrote this rather quickly, perhaps it can be improved.